File: cpp/211_Add_and_Search_Word_Data_structure_design.cpp
/*
Design a data structure that supports the following two operations:

void addWord(word)
bool search(word)

search(word) can search a literal word or a regular expression string containing only letters a-z or .. A . means it can represent any one letter.

Example:

addWord("bad")
addWord("dad")
addWord("mad")
search("pad") -> false
search("bad") -> true
search(".ad") -> true
search("b..") -> true

Note:
You may assume that all words are consist of lowercase letters a-z.
*/
#include <iostream>

using namespace std;

class WordNode {
  public: 
    bool isWord = false;
    WordNode* next[26]{};
    WordNode* get(char c) {
      return next[c - 'a'];
    }
    void add(char c) {
      next[c - 'a'] = new WordNode();
    }
};

class WordDictionary {
  public:
    /** Initialize your data structure here. */
    WordDictionary() {
      
    }
    
    /** Adds a word into the data structure. */
    void addWord(string word) {
      WordNode* w = root;
      for(auto c : word) {
        if(!w->get(c)) w->add(c);
        w = w->get(c);
      }
      w->isWord = true;
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
      return find(word, 0, root);
    }

  private:
    WordNode* root = new WordNode();
    bool find(string& word, int pos, WordNode* root) {
      if(pos == word.size()) return root->isWord;
      if(word[pos] != '.') {
        root = root->get(word[pos]);
        return root ? find(word, pos + 1, root) : false;
      }
      for(int i = 0; i < 26; i++) {
        if(root->next[i] && find(word, pos + 1, root->next[i]))
          return true;
      }
      return false;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * bool param_2 = obj.search(word);
 */

File: py/211_Add_and_Search_Word_Data_structure_design.py
import collections

class TrieNode():
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.isWord = False

class WordDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        """
        Adds a word into the data structure.
        """
        node = self.root
        for w in word:
            node = node.children[w]
        node.isWord = True

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        """
        node = self.root
        self.isMatch = False
        self.dfs(node, word)
        return self.isMatch

    def dfs(self, node: TrieNode, word: str):
        if not word:
            if node.isWord:
                self.isMatch = True
            return
        if word[0] == '.':
            for n in node.children.values():
                self.dfs(n, word[1:])
        else:
            node = node.children.get(word[0])
            if not node:
                return
            self.dfs(node, word[1:])



sol = WordDictionary()
sol.addWord('asd')
sol.addWord('as')
sol.addWord('tree')
# for i in sol.root.children:
#     if i.children:
#         for j in i.children:
#             print(j)
#     else:
#         print('root -- ', i)
